src/flask_session/dynamodb/dynamodb.py

```diff
@@ -0,0 +1,126 @@
+from decimal import Decimal
+from typing import Optional
+
+import boto3
+from mypy_boto3_dynamodb.service_resource import DynamoDBServiceResource
+from flask import Flask
+from itsdangerous import want_bytes
+
+from ..base import ServerSideSession, ServerSideSessionInterface
+from ..defaults import Defaults
+
+
+class DynamoDBSession(ServerSideSession):
+    pass
+
+
+class DynamoDBSessionInterface(ServerSideSessionInterface):
+    """A Session interface that uses dynamodb as backend. (`boto3` required)
+
+    :param client: A ``DynamoDBServiceResource`` instance.
+    :param key_prefix: A prefix that is added to all DynamoDB store keys.
+    :param use_signer: Whether to sign the session id cookie or not.
+    :param permanent: Whether to use permanent session or not.
+    :param sid_length: The length of the generated session id in bytes.
+    :param table_name: DynamoDB table name to store the session.
+
+    .. versionadded:: 0.6
+        The `sid_length` parameter was added.
+
+    .. versionadded:: 0.2
+        The `use_signer` parameter was added.
+    """
+
+    session_class = DynamoDBSession
+
+    def __init__(
+        self,
+        app: Flask,
+        client: Optional[DynamoDBServiceResource] = Defaults.SESSION_DYNAMODB,
+        key_prefix: str = Defaults.SESSION_KEY_PREFIX,
+        use_signer: bool = Defaults.SESSION_USE_SIGNER,
+        permanent: bool = Defaults.SESSION_PERMANENT,
+        sid_length: int = Defaults.SESSION_ID_LENGTH,
+        serialization_format: str = Defaults.SESSION_SERIALIZATION_FORMAT,
+        table_name: str = Defaults.SESSION_DYNAMODB_TABLE,
+    ):
+
+        if client is None:
+            warnings.warn(
+                "No valid DynamoDBServiceResource instance provided, attempting to create a new instance on localhost:8000.",
+                RuntimeWarning,
+                stacklevel=1,
+            )
+            client = boto3.resource(
+                "dynamodb",
+                endpoint_url="http://localhost:8000",
+                region_name="us-west-2",
+                aws_access_key_id="dummy",
+                aws_secret_access_key="dummy",
+            )
+
+        try:
+            client.create_table(
+                AttributeDefinitions=[
+                    {"AttributeName": "id", "AttributeType": "S"},
+                ],
+                TableName=table_name,
+                KeySchema=[
+                    {"AttributeName": "id", "KeyType": "HASH"},
+                ],
+                BillingMode="PAY_PER_REQUEST",
+            )
+            client.meta.client.get_waiter("table_exists").wait(TableName=table_name)
+            client.meta.client.update_time_to_live(
+                TableName=self.table_name,
+                TimeToLiveSpecification={
+                    "Enabled": True,
+                    "AttributeName": "expiration",
+                },
+            )
+        except (AttributeError, client.meta.client.exceptions.ResourceInUseException):
```

Comment:
I think there may be a problem here:

Scenario:
- Running inside a real environment, trying to use existing table.
- The instance or whatever is running this code doesn't have `dynamodb:CreateTable` or `dynamodb:UpdateTimeToLive` or `dynamodb:Describe/GetTable` permissions
- I'd expect an access denied error in this case, because it doesn't even get to the point of checking if the table exists etc.
- This except probably doesn't handle that.

This means we'd have to give whichever resource is running this permission to use the CreateTable etc. APIs even though we want to use an existing table. Not ideal from a security perspective.

(Purely theoretical at this point, but I'm fairly certain this is what would happen)

Reply:
Fair point, seems more reasonable to check if the table exists before the try, is that what you would suggest? Could you add a PR to development?

Reply:
From a security perspective I'd prefer that the implementation can work with minimal permissions, in this case:

- `dynamodb:PutItem`
- `dynamodb:GetItem`

This would be sufficient for read and write access to a properly configured table.

All this table and TTL management is something that we usually like to handle through Infrastructure as Code. Many companies also require additional configuration regarding backups, etc.

Having the option of doing that through flask-session is nice for convenience, but IMO a bit too much magic that happens by default.

I'd prefer an `create_table` flag or something like that, which is should be disabled by default.
If we find a good approach that balances different requirements I may be able to add a PR.

Reply:
I see what you mean. I think the first step would be to factor out the creation into a single method on each interface.  Similar to what is done here [https://github.com/kroketio/quart-session/blob/master/quart_session/sessions.py](https://github.com/kroketio/quart-session/blob/master/quart_session/sessions.py).This could then be easily subclassed to skip the method (maybe like 3 lines of code).

The second step would be adding the config setting for all interfaces, which I am less sure of (adding more settings) unless there is a bit of demand or discussing with the team.

I think if the first was done neatly it would be merged quickly.

Reply:
I'm currently looking into this, but I'm not super familiar with the codebase yet.

My approach would be to:

1. Refactor `DynamoDBSessionInterface` to have the create table logic as a separate function
2. Create a subclass `ExistingDynamoDBTableSessionInterface` that skips the create logic
3. Add a `SESSION_TYPE` `existing_dynamodb_table` to `flask_session/__init__.py` that uses 2)
4. Tests, obviously

This would result in there being two DynamoDB varieties for the `SESSION_TYPE`.
That would make this implementation different from the others.

I'm considering another optional flag to the constructor of `DynamoDBSessionInterface`, something like `table_already_exists` (Default: False), which would skip the `create` function.
This would probably require another DynamoDB parameter, e.g. `SESSION_DYNAMODB_TABLE_ALREADY_EXISTS`.

Not sure what is a better approach, any thoughts @Lxstr ?

Reply:
I think we can simplify a bit

Step 1 is good.

This is already done for postgresql (impending PR) but would also need to be done for SQLAlchemy and the base class

Steps 2 and 3 can be instead adding the flag SESSION_TABLE_EXISTS

Base class should look something like:
```
class ServerSideSessionInterface(FlaskSessionInterface, ABC):
    ...

    def __init__(... session_table_exists: Optional[bool] = Defaults.SESSION_TABLE_EXISTS,):
        ...
        if not session_table_exists:
            self._create_schema_and_table()

    def _create_schema_and_table():
        pass
```

Tests would be great

Does that make sense?

Reply:
Thanks, looks good to me, I've added a variation of that as a PR: https://github.com/pallets-eco/flask-session/pull/237